app/controllers/samples_controller.rb

```diff
@@ -233,15 +233,38 @@ def report_info
 
     @report_info = external_report_info(pipeline_run_id, background_id, params)
 
+    # Get list of tax_ids to look up in TaxonLineage and TaxonCount rows. Include family_taxids.
     tax_ids = @report_info[:taxonomy_details][2].map { |x| x['tax_id'] }
+    tax_ids = tax_ids | @report_info[:taxonomy_details][2].map { |x| x['family_taxid'] }
     lineages = TaxonCount.connection.select_all(TaxonLineage.where(taxid: tax_ids)).to_hash
     lineage_by_taxid = {}
     lineages.each do |x|
       lineage_by_taxid[x['taxid']] = x
     end
 
+    # Set lineage info from the first positive tax_id of the species, genus, or family levels.
+    # Preserve names of the negative 'non-specific' nodes.
     @report_info[:taxonomy_details][2].each do |tax|
-      tax['lineage'] = lineage_by_taxid[tax['tax_id']]
+      if tax['tax_id'] > 0
+        tax['lineage'] = lineage_by_taxid[tax['tax_id']]
+      elsif tax['tax_id'] < 0 && tax['tax_level'] == 1 && tax['genus_taxid'] > 0
+        # NOTE: start_tax_lineage fills missing info from TaxonLineage with an empty hash.
+        tax['lineage'] = lineage_by_taxid[tax['genus_taxid']] || {}
+        tax['lineage']['taxid'] = tax['tax_id']
+        tax['lineage']['species_taxid'] = tax['species_taxid']
+        tax['lineage']['species_name'] = tax['name']
+      elsif tax['tax_id'] < 0 && tax['tax_level'] == 1 && tax['genus_taxid'] < 0 && tax['family_taxid'] > 0
+        tax['lineage'] = lineage_by_taxid[tax['family_taxid']] || {}
+        tax['lineage']['taxid'] = tax['tax_id']
+        tax['lineage']['species_taxid'] = tax['species_taxid']
+        tax['lineage']['species_name'] = tax['name']
+        tax['lineage']['genus_taxid'] = tax['genus_taxid']
+      elsif tax['tax_id'] < 0 && tax['tax_level'] == 2 && tax['family_taxid'] > 0
+        tax['lineage'] = lineage_by_taxid[tax['family_taxid']] || {}
+        tax['lineage']['taxid'] = tax['tax_id']
+        tax['lineage']['genus_taxid'] = tax['genus_taxid']
+        tax['lineage']['genus_name'] = tax['name']
+      end
```

Comment:
This looks too complicated.   Shouldn't it just be as follows:
```
    def most_specific_positive_id(tax):
            [tax['species_taxid'], tax['genus_taxid'], tax['family_taxid']].each |tentative_id| do
            if tentative_id > 0:
                return tentative_id
            return 0

lineage_id = most_specific_positive_id(tax)
if lineage_id
    lineage = lineage_by_taxid[lineage_id]
else
    lineage = [-100, -200, ..., -700]
lineage['species_taxid'] = tax['species_taxid']
lineage['genus_taxid'] = tax['genus_taxid']
lineage['family_taxid'] = tax['family_taxid']
```

Reply:
Yes, it would be great to simplify in a similar way. However, it is crucial that we also set `tax['lineage']['#{level_name (tax['tax_level'])}_name'] = tax['name']`

Reply:
Also note that in the very near future, after migration, this should change to simply copy the lineage from the taxon.   Or better yet change JS to use the lineage from the taxon.   It's only a temporary situation that the taxon here doesn't have the complete lineage.
```
lineage['species_taxid'] = tax['species_taxid']
lineage['genus_taxid'] = tax['genus_taxid']
lineage['family_taxid'] = tax['family_taxid']
...
```

Reply:
What do you mean by "doesn't seem to fill out all the lineages"? We nee lineage_by_taxid to fill out the part of the lineage that is above family level, and make sure to keep the info from the taxon_counts to fill the lineages below or at family level

Reply:
Oh my bad I missed completely that we were also setting the names not just the ids here.

Reply:
Yeah I like the line 
```
tax['lineage']["#{level_name(tax['tax_level'])}_name"] = tax['name']
```
Is that the only name that needs to be passed up?   I am actually curious how the tree viz makes up the higher category names --- does it get them from here, or does it get them from the lineage map?  If the latter, it will need the modulus arithmetic there.   What might be most efficient, of course, is to return to it the subset of the lineage tree that it needs, but that may be in the scope of another pull request.

Reply:
Right, so my recommendation is to get rid of those if statements and simplify as suggested, including Charles' way to copy the name at the tax level, since that's easier to maintain (in future we just delete the lineage lookup and all comes from that tax info).

Reply:
The higher rank names come from lineage_by_taxid, which is directly derived from TaxonLineage. 
`tax` here is derived from the same hash that corresponds to a row in the report: it contains only the name of its own level.

Reply:
Thanks.  I also realized we may need to handle the case then the lineage info doesn't exist for the most specific taxid.   In those cases we still have the specices, genus, family from the tax hash (ultimately from pipeline) but nothing higher up.  Also the superkingdom won't be available.  Probably shouldn't worry too much about such legitimately uncategorized things though;  as they are also supposed to disappear if lineage is updated in sql (because should only happen if pipeline lineage is ahead of webapp lineage).